pkg/dataplane/identity.go

```diff
@@ -1,7 +1,133 @@
+	azcloud "github.com/Azure/azure-sdk-for-go/sdk/azcore/cloud"
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/msi-dataplane/internal/swagger"
+	"golang.org/x/crypto/pkcs12"
+)
+
+var (
+	// Errors returned when processing idenities
+	errDecodeClientSecret = errors.New("failed to decode client secret")
+	errParseCertificate   = errors.New("failed to parse certificate")
+	errNilField           = errors.New("expected non nil field in identity")
+	errNotRSAKey          = errors.New("pkcs#12 certificate must contain a RSA private key")
+	errResourceIDNotFound = errors.New("resource ID not found in user-assigned managed identity	")
+)
+
+// CredentialsObject is a wrapper around the swagger.CredentialsObject to add additional functionality
+// swagger.Credentials object can represent either system or user-assigned managed identity
 type CredentialsObject struct {
 	swagger.CredentialsObject
+	cloud string
+}
+
+type UserAssignedIdentities struct {
+	CredentialsObject
+}
+
+// This method may be used by clients to check if they can use the object as a user-assigned managed identity
+// Ex: get credentials object from key vault store and check if it is a user-assigned managed identity to call client for object refresh.
+func (c CredentialsObject) IsUserAssigned() bool {
+	return len(c.ExplicitIdentities) > 0
+}
+
+func (u UserAssignedIdentities) GetCredential(resourceID string) (*azidentity.ClientCertificateCredential, error) {
+	for _, id := range u.ExplicitIdentities {
+		if id != nil && id.ResourceID != nil {
+			if *id.ResourceID == resourceID {
+				return getClientCertificateCredential(*id, u.cloud)
+			}
+		}
+	}
+
+	return nil, errResourceIDNotFound
+}
+
+func getClientCertificateCredential(identity swagger.NestedCredentialsObject, cloud string) (*azidentity.ClientCertificateCredential, error) {
+	// Double check nil pointers so we don't panic
+	if identity.ClientID == nil {
+		return nil, fmt.Errorf("%w: clientID", errNilField)
+	}
+	if identity.TenantID == nil {
+		return nil, fmt.Errorf("%w: tenantID", errNilField)
+	}
+	if identity.ClientSecret == nil {
+		return nil, fmt.Errorf("%w: clientSecret", errNilField)
+	}
+	if identity.AuthenticationEndpoint == nil {
+		return nil, fmt.Errorf("%w: authenticationEndpoint", errNilField)
+	}
+
+	// Set the regional AAD endpoint
+	// https://eng.ms/docs/products/arm/rbac/managed_identities/msionboardingcredentialapiversion2019-08-31
+	opts := &azidentity.ClientCertificateCredentialOptions{
+		ClientOptions: azcore.ClientOptions{
+			Cloud: getAzCoreCloud(cloud),
+		},
+	}
+	opts.Cloud.ActiveDirectoryAuthorityHost = *identity.AuthenticationEndpoint
+
+	// Parse the certificate and private key from the base64 encoded secret
+	decodedSecret, err := base64.StdEncoding.DecodeString(*identity.ClientSecret)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %w", errDecodeClientSecret, err)
+	}
+	key, crt, err := pkcs12.Decode(decodedSecret, "")
```

Comment:
We could use az identity's function: https://github.com/Azure/azure-sdk-for-go/blob/48d39a82091b3aebe3df3505bd5372294b3461ed/sdk/azidentity/client_certificate_credential.go#L94

I don't b/c MSI team says to treat certs like a black box, and that function doesn't support SHA256 MAC so I'd rather have better compat than reusing azidentity.

Reply:
Discussed this offline - we'll use azidentity for now and revert this change if we encounter issues during canary testing. Added code comment to highlight this issue.